#**Laptop Dataset**

###This dataset contains information about laptops such as Brand, Price, Processor, RAM, Operating System, etc. It can be used to gain insights into laptop specifications and pricing trends, and to build machine learning models for tasks like price prediction.

In [1]:
# Importing dependencies
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from scipy import stats
import re

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/nikhil-xyz/datasets/refs/heads/main/laptop_uncleaned.csv')

In [3]:
df.shape

(2843, 17)

In [4]:
df.isna().sum()

Title                  0
Brand                 33
Screen_Size           87
CPU_Model            215
Ram                  132
Operating_System     120
Rating               560
Series               342
Color                734
Resolution          2661
Weight              2777
USB                 2808
Battery             2836
Graphics             780
Disk_size            595
Card_desc            208
Price                980
dtype: int64

### Checking whether dataset contains any duplicate entries and removing them if present.

In [5]:
df.duplicated().all().sum()

0

##**Handling missing values**

##**Addressing Brand**

In [6]:
df.Brand.unique()

array(['Microsoft', 'HP', 'Dell', 'MSI', 'Alienware', 'EXCaliberPC',
       'GIGABYTE', 'Panasonic', 'Toughbook', 'Apple', 'Ficihp',
       'Packard Bell', 'B BELK', 'Acer', 'Lenovo', 'VGKE', 'Jumper',
       'ASUS', 'BiTECOOL', nan, 'CHUWI', 'LG', 'WAKST', 'ECOHERO',
       'Fusion5', 'TECLAST', 'Panasonic Toughbook CF-31 Rugged', 'Bmax',
       'LANRUO', 'SAMSUNG', 'Blackview', 'Sceptre',
       'HP-Consumer Remarketing', 'HYUNDAI', 'Fullcom Tech', 'Generic',
       'SewSwank', 'Broage', 'Razer', 'LincPlus', 'HP Tuners', 'Unknown',
       'Laptop', 'Best Notebooks', 'DERE',
       'REPOWER REPAIR REFURBISH REUSE', 'Artist Unknown', 'Miwasion',
       'Carlisle Paddles', 'OTVOC', 'Tomtoc', 'Thomson', 'Coolby'],
      dtype=object)

In [7]:
df.Brand.isna().sum()

33

In [8]:
brand_missing = df[df.Brand.isna()]

In [9]:
brand_missing.Title.values

array(['Acer Aspire 1 A115-32-C96U Laptop | 15.6" FHD | Celeron N4500 | 4GB DDR4 | 128GB eMMC | MS 365 Personal 1-Year Subscription | Win 11 Home in S Mode with AOPEN 24SA2Y bi 23.8" FHD VA-Monitor',
       'Acer Aspire 1 A115-32-C96U Laptop | 15.6" FHD | Celeron N4500 | 4GB DDR4 | 128GB eMMC | WiFi 5 | MS 365 Personal 1-Year Subscription | Win 11 Home in S Mode USB Type-C Dock D501, 60W PD',
       'PANASONIC FZ-VZSU88U',
       'Panasonic Toughbook CF-52 500GB 4GB 15.4" Genuine Win 7 Pro + Microsoft Office 2010',
       'Acer Aspire 3 A315-24P-R2SC Laptop | 15.6" FHD IPS Display | AMD Ryzen 5 7520U | Radeon Graphics | 8GB LPDDR5 | 256GB SSD | Wi-Fi 5 | Windows 11 Home USB 3.0 Dock U301',
       'Acer Aspire 3 A314-23P-R3QA Laptop | 14.0" FHD IPS | AMD Ryzen 5 7520U | AMD Radeon Graphics | 8GB LPDDR5 | 512GB SSD | Wi-Fi 6 | Windows 11 Home FHD USB 2MP Webcam with Digital Microphone',
       'Acer Aspire 3 A315-24PT-R08Z Laptop | 15.6" FHD IPS Touch | AMD Ryzen 3 7320U | Radeon Graphic

###**Observations and Modifications**
- Brand value for all the missing entries is present inside the Title feature as a first string
- Simple lambda function will be able to retrieve it.

In [10]:
brand_missing['Brand'] = brand_missing.Title.apply(lambda x: x.split()[0])

C:\Program Files\KMSpico\temp\ipykernel_10308\862350388.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  brand_missing['Brand'] = brand_missing.Title.apply(lambda x: x.split()[0])


In [11]:
# Replacing all the missing entries
df.Brand.fillna(brand_missing.Brand, inplace=True)

In [12]:
df.isna().sum()

Title                  0
Brand                  0
Screen_Size           87
CPU_Model            215
Ram                  132
Operating_System     120
Rating               560
Series               342
Color                734
Resolution          2661
Weight              2777
USB                 2808
Battery             2836
Graphics             780
Disk_size            595
Card_desc            208
Price                980
dtype: int64

##**Addressing Screen_Size**

In [13]:
# Filtering out missing rows
screen_missing = df[df.Screen_Size.isna()]

In [14]:
screen_missing.Title.values

array(['B BELK Compatible with MacBook Air 13 inch Case M1, MacBook Air Case 2021 2020 2019 2018 Model A2337 A2179 A1932 with Touch ID, Clear Plastic Laptop Hard Shell + 2 Keyboard Covers + Screen Protector',
       'Apple 2022 MacBook Air Laptop with M2 chip: 13.6-inch Liquid Retina Display, 8GB RAM, 256GB SSD Storage, Backlit Keyboard, 1080p FaceTime HD Camera. Works with iPhone and iPad; Midnight AppleCare',
       'Apple MacBook Pro 16" with Liquid Retina XDR Display, M1 Pro Chip with 10-Core CPU and 16-Core GPU, 32GB Memory, 1TB SSD, Silver, Late 2021',
       'Apple 2022 MacBook Air Laptop with M2 chip: 13.6-inch Liquid Retina Display, 8GB RAM, 256GB SSD Storage, Backlit Keyboard, 1080p FaceTime HD Camera. Works with iPhone and iPad; Starlight AppleCare',
       'Apple 2021 MacBook Pro (14-inch, M1 Pro chip with 10‑core CPU and 16‑core GPU, 16GB RAM, 1TB SSD) - Space Gray AppleCare+ for 14-inch MacBook Pro',
       'Apple 2022 MacBook Air Laptop with M2 chip: 13.6-inch Liquid Ret

In [15]:
screen_missing.shape

(87, 17)

###**Observations and Modifications**
- Most of the missing screen sizes are available inside the Title section
- Values can be retrieved using regular expression

**Steps:-**

- Filtering out all the missing entries and store it inside new dataframe
- Defining new function which extracts screen size using regular expression from the title section of the screen_missing dataframe
- Once retrieval completes, merge updated datarfame with original one.
- Drop entries which still have missing values.

In [16]:
# Defining regular expression and retrieving values
screen_missing['Screen_Size'] = screen_missing.Title.str.extract(r'(\d{2}\.\d{1}|\d{2})(?:\"|\sinch|\-inch|\”|\sInch)')
screen_missing['Screen_Size'].isna().sum()


C:\Program Files\KMSpico\temp\ipykernel_10308\600362279.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  screen_missing['Screen_Size'] = screen_missing.Title.str.extract(r'(\d{2}\.\d{1}|\d{2})(?:\"|\sinch|\-inch|\”|\sInch)')


23

In [17]:
screen_missing.Screen_Size.unique()

array(['13', '13.6', '16', '14', '13.3', '15.6', nan, '14.0', '15.4',
       '11.6', '17.3', '15', '13.5', '14.4', '12.4'], dtype=object)

In [18]:
# Merging screen_missing with an original dataframe
df.Screen_Size.fillna(screen_missing.Screen_Size, inplace=True)

In [19]:
df.Screen_Size.unique()

array(['13.5 Inches', '13 Inches', '12.4 Inches', '14 Inches',
       '34 Inches', '15.6 Inches', '17.3 Inches', '17 Inches',
       '13.4 Inches', '16 Inches', '15.6', '15 Inches', '15.4 Inches',
       '13', '13.3 Inches', '13.6', '16.2 Inches', '16', '14', '14.2',
       '10.9 Inches', '9.7 Inches', '13.3', '14.1 Inches', '12.3 Inches',
       '11.6 Inches', '13.1 Inches', '12 Inches', '11 Inches',
       '12.5 Inches', nan, '10.1 Inches', '11.6', '17.3', '14.0', '15.4',
       '10.4 Inches', '31.5 Inches', '6 Inches', '5.6 Inches',
       '10 Inches', '17', '12.45 Inches', '30 Inches', '15',
       '12.2 Inches', '24 Inches', '14.2 Inches', '13.6 Inches',
       '18.4 Inches', '1 Centimeters', '17.25 Inches', '0.01',
       '0.1 Inches', '14.5', '12.1 Inches', '14.5 Inches', '10.5 Inches',
       '11', '116 Inches', '16.6 Inches', '3 Inches', '13.5',
       '14.4 Inches', '14.4', '12.4', '10.51 Inches'], dtype=object)

###**Observations**
- We have successfully extracted screen size from the Title for the missing entries.
- But original entries where screen size was not missing values are given with screen size and its unit of measure.
- We need to conduct further preprocessing to remove unit of measure from such entries
- Regular expressions can be utilized to achieve that.

In [20]:
# Defining regular expression and retrieving values
df['Screen_Size'] = df.Screen_Size.str.extract(r'(\d{1,2}\.\d{1,2}|\d{1,2})')

In [21]:
df.Screen_Size.unique()

array(['13.5', '13', '12.4', '14', '34', '15.6', '17.3', '17', '13.4',
       '16', '15', '15.4', '13.3', '13.6', '16.2', '14.2', '10.9', '9.7',
       '14.1', '12.3', '11.6', '13.1', '12', '11', '12.5', nan, '10.1',
       '14.0', '10.4', '31.5', '6', '5.6', '10', '12.45', '30', '12.2',
       '24', '18.4', '1', '17.25', '0.01', '0.1', '14.5', '12.1', '10.5',
       '16.6', '3', '14.4', '10.51'], dtype=object)

In [22]:
# Changing datatype from object to float
df.Screen_Size = df.Screen_Size.astype(float)

In [23]:
df.Screen_Size.value_counts().sort_index()

Screen_Size
0.01       1
0.10       1
1.00       4
3.00       1
5.60       1
6.00       1
9.70       5
10.00      2
10.10     25
10.40      4
10.50      3
10.51      1
10.90      1
11.00     11
11.60    136
12.00     18
12.10      1
12.20     12
12.30     14
12.40     32
12.45      3
12.50     11
13.00     48
13.10     40
13.30    153
13.40     18
13.50     62
13.60      6
14.00    593
14.10     15
14.20      3
14.40     10
14.50      5
15.00     76
15.40     20
15.60    953
16.00    137
16.20      4
16.60      1
17.00     46
17.25      1
17.30    330
18.40      6
24.00      1
30.00      1
31.50      2
34.00      1
Name: count, dtype: int64

In [24]:
df[df.Screen_Size < 10][['Title', 'Screen_Size', 'Price']].values

array([['Apple iPad Air 2, 16 GB, Silver, Newest Version (Renewed)', 9.7,
        nan],
       ['2014 Apple iPad Air 2 (9.4 inch, Wi-Fi, 64GB) Silver (Renewed)',
        9.7, nan],
       ['Apple iPad Air 2, 16 GB, Gold, (Renewed)', 9.7, nan],
       ['Apple iPad Air 2, 16 GB, Space Gray (Renewed)', 9.7, nan],
       ['[512GB M.2 SSD Version] GPD MicroPC Micro PC 6-inch Handheld Industry Laptop Mini PC Computer Win 10 Pro Supports RS-232,RJ45,Type C,USB 3.0,8GB RAM',
        6.0, nan],
       ['Panasonic Toughbook U1 - Atom Z520 1.33 GHz - 5.6" TFT (CR4302) Category: Laptop Computers',
        5.6, nan],
       ['Apple iPad Air 2, 16 GB, Space Gray (Renewed)', 9.7, nan],
       ['Alienware Aurora R11 Gaming Desktop 10th Gen Intel Core i7, 16GB Dual Channel HyperX Fury DDR4 XMP, NVIDIA GeForce RTX 3080, 1TB HDD + 512GB SSD, Killer Wi-Fi 6, Black (Latest Model)',
        1.0, '$3,589.99'],
       ['Alienware Aurora R10 Gaming Desktop - AMD Ryzen 9 5900, 32GB DDR4 RAM, 1TB SSD, NVIDIA GeF

### There are some entries in screen size which can be further reassigned after preprocessing the Price feature. For a time being, we will drop all entries that couldn't retrieve any information from the Title section.

In [25]:
df = df.dropna(subset = ['Screen_Size'], axis=0)

In [26]:
df.isna().sum()

Title                  0
Brand                  0
Screen_Size            0
CPU_Model            203
Ram                  122
Operating_System     107
Rating               550
Series               331
Color                718
Resolution          2640
Weight              2756
USB                 2785
Battery             2813
Graphics             763
Disk_size            575
Card_desc            191
Price                969
dtype: int64

##**Addressing CPU_Model**

In [27]:
df[~df['CPU_Model'].isna()]

Title      Brand  \
0     Microsoft Surface Laptop Intel Core i5 7th Gen...  Microsoft   
1     Microsoft Surface Laptop 3 13.5" Touch-Screen ...  Microsoft   
2     Microsoft Surface Laptop (Intel Core i7, 16GB ...  Microsoft   
3     Microsoft Surface Laptop 3, 10th Gen Intel Cor...  Microsoft   
4     2022 Microsoft Surface Laptop Go 12.4" Touchsc...  Microsoft   
...                                                 ...        ...   
2838  Microsoft Surface Laptop Go 12.4" Touchscreen ...  Microsoft   
2839  Microsoft Surface Laptop Go 12.4" Touchscreen ...  Microsoft   
2840  Microsoft Surface Laptop Go 12.4in Touchscreen...  Microsoft   
2841  Microsoft Surface Laptop Go 12.4" Touchscreen ...  Microsoft   
2842  Microsoft Surface Laptop Go 12.4" Touchscreen ...  Microsoft   

      Screen_Size                   CPU_Model    Ram Operating_System  Rating  \
0            13.5  1.2GHz Cortex A8 Processor   8 GB       Windows 10     4.4   
1            13.5                     Core i5   8 GB       Windows 10     4.4   
2            13.5                     Core i7  16 GB     Windows 10 S     3.9   
3            13.0              Core i5-1035G7   8 GB  Windows 10 Home     4.2   
4            12.4                     Core i5   4 GB   Windows 11 Pro     3.9   
...           ...                         ...    ...              ...     ...   
2838         12.4              Core i5-1035G1   4 GB   Windows 10 Pro     4.2   
2839         12.4              Core i5-1035G1   4 GB   Windows 10 Pro     4.2   
2840         12.4              Core i5-1035G1   8 GB       Windows 10     4.5   
2841         12.4              Core i5-1035G1   4 GB   Windows 10 Pro     4.2   
2842         12.4              Core i5-1035G1   4 GB  Windows 10 Home     4.4   

                 Series        Color Resolution Weight  USB Battery  \
0                 Intel     Platinum        NaN    NaN  NaN     NaN   
1     Microsoft Surface  Matte Black        NaN    NaN  NaN     NaN   
2        Surface Laptop     Platinum        NaN    NaN  NaN     NaN   
3      Surface Laptop 3  Cobalt Blue        NaN    NaN  NaN     NaN   
4     Surface Laptop Go       Silver        NaN    NaN  NaN     NaN   
...                 ...          ...        ...    ...  ...     ...   
2838         Surface Go     Platinum        NaN    NaN  NaN     NaN   
2839         Surface Go     Platinum        NaN    NaN  NaN     NaN   
2840  Surface Laptop Go     Ice Blue        NaN    NaN  NaN     NaN   
2841         Surface Go     Platinum        NaN    NaN  NaN     NaN   
2842       MS1ZO00001OB     Platinum        NaN    NaN  NaN     NaN   

                          Graphics Disk_size   Card_desc    Price  
0            Intel HD Graphics 620    256 GB  Integrated      NaN  
1                              NaN       NaN         NaN      NaN  
2     Intel Iris Plus Graphics 640    512 GB  Integrated      NaN  
3                  Intel Iris Plus       NaN  Integrated      NaN  
4                              NaN    256 GB  Integrated  $359.99  
...                            ...       ...         ...      ...  
2838            Intel UHD Graphics     64 GB  Integrated  $262.99  
2839            Intel UHD Graphics     64 GB  Integrated  $262.99  
2840            Intel UHD Graphics    128 GB  Integrated      NaN  
2841            Intel UHD Graphics     64 GB  Integrated  $262.99  
2842                           NaN     64 GB  Integrated  $254.99  

[2617 rows x 17 columns]

In [28]:
df.CPU_Model.unique()

array(['1.2GHz Cortex A8 Processor', 'Core i5', 'Core i7',
       'Core i5-1035G7', 'Ryzen 3 2300U', 'Intel Core i7-11700',
       'Intel Core i5-1135G7', 'Ryzen 5', 'Intel Core i9',
       'Core i7 8750H', 'Intel Core i7', 'AMD Ryzen 9 5900X',
       'Core i7 6600U', 'Ryzen 9', 'Core i9', 'Corei5-10300H',
       'Core i7-3720QM', nan, 'Celeron N4020', 'Core i7 Family',
       'Intel Core i5', 'Ryzen 7', 'ARM 7100', 'Apple M1',
       'Core i5 8250U', 'Unknown', 'Core i5-6267U', 'Core i3',
       'Core I7 11800H', 'Corei7-10750H', 'AMD Ryzen 7', 'Core i5 Family',
       'Corei7-9750H', 'Ryzen 5 5500U', 'Core i7-10510U',
       'Core i3 Family', 'Corei7-1065G7', 'Xeon Platinum 8160M',
       'Core I9 12900H', 'AMD Ryzen 5 5500U', 'Celeron', 'Others',
       'Celeron N', 'AMD A Series', 'Celeron N4000', 'Core i5-5250U',
       'MediaTek MT8183', 'Ryzen 3 3350U', 'Core i7 Extreme', 'AMD A6',
       'Core i5-3340s', '8032', 'Pentium', 'Core I3 1115G4',
       'Intel Core i3', 'Cortex', 'Ry

###**Observations and Modifications**
- It looks like majority of models belongs to Intel's Core i series, Ryzen series, Celeron series.
- As usual, model information is also mention inside the Title section

**Steps:-**
- Create new dataframe containing all the missing values for the CPU_Model section
- Define a function which can retrieve model infoemation from the Title section using regular expressions
- Regular expression will be design to retrieve model from Intel's Core i, Ryzen, Celeron series.
- After retrieval, merge new dataframe with the original one.

In [29]:
cpu_missing = df[df.CPU_Model.isna()]

In [30]:
cpu_pattern = r'(Intel\s*i\d|Core\s*i\d|Ryzen\s*\d|Radeon\s*R\d|Exynos\s*\d|Ultra\s*\d)'
cpu_missing['CPU_Model'] = cpu_missing.Title.str.extract(cpu_pattern)

C:\Program Files\KMSpico\temp\ipykernel_10308\3171287869.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_missing['CPU_Model'] = cpu_missing.Title.str.extract(cpu_pattern)


In [31]:
cpu_missing['CPU_Model'].isna().sum()

124

In [32]:
cpu_missing[cpu_missing['CPU_Model'].isna()]['Title'].values

array(['Ficihp 13.3" Laptop Screen Extender for M1 MacBook One Cable Plug&Play, Cooling System Triple Portable Monitor for Laptop with USB C HDMI, Monitor Attachment for Laptop Frame 13-16 inch Mac/Android',
       'B BELK Compatible with MacBook Air 13 inch Case M1, MacBook Air Case 2021 2020 2019 2018 Model A2337 A2179 A1932 with Touch ID, Clear Plastic Laptop Hard Shell + 2 Keyboard Covers + Screen Protector',
       'Apple 2022 MacBook Air Laptop with M2 chip: 13.6-inch Liquid Retina Display, 8GB RAM, 256GB SSD Storage, Backlit Keyboard, 1080p FaceTime HD Camera. Works with iPhone and iPad; Midnight AppleCare',
       'Apple MacBook Pro 13.3" with Retina Display, M1 Chip with 8-Core CPU and 8-Core GPU, 16GB Memory, 512GB SSD, Silver, Late 2020',
       'Apple MacBook Pro 16" with Liquid Retina XDR Display, M1 Pro Chip with 10-Core CPU and 16-Core GPU, 32GB Memory, 2TB SSD, Space Gray, Late 2021',
       'Apple MacBook Pro 16" with Liquid Retina XDR Display, M1 Pro Chip with 10-Core

In [33]:
df.isna().sum()

Title                  0
Brand                  0
Screen_Size            0
CPU_Model            203
Ram                  122
Operating_System     107
Rating               550
Series               331
Color                718
Resolution          2640
Weight              2756
USB                 2785
Battery             2813
Graphics             763
Disk_size            575
Card_desc            191
Price                969
dtype: int64

###There are still some missing values left to fill. We will address missing entries of CPU_Model futher once we finish preprocessing the Price feature.

##**Addressing Price Feature**

In [34]:
df.Price.unique()

array([nan, '$359.99', '$2,167.00', '$818.00', '$699.00', '$2,449.99',
       '$1,799.95', '$1,699.00', '$3,049.00', '$1,899.00', '$1,549.00',
       '$4,345.00', '$1,799.00', '$3,295.00', '$1,399.00', '$1,949.99',
       '$689.99', '$2,499.00', '$2,649.00', '$3,199.00', '$4,595.00',
       '$5,195.00', '$2,319.00', '$799.00', '$599.99', '$229.99',
       '$12.86', '$978.75', '$1,658.52', '$1,099.00', '$2,399.00',
       '$2,799.00', '$3,599.00', '$1,449.00', '$1,199.00', '$1,499.00',
       '$1,619.00', '$2,899.00', '$1,299.00', '$854.21', '$1,579.00',
       '$1,468.95', '$1,242.96', '$950.61', '$1,649.00', '$2,084.45',
       '$2,699.00', '$239.00', '$2,999.00', '$870.00', '$1,536.99',
       '$1,045.52', '$714.49', '$2,409.00', '$1,496.24', '$1,749.99',
       '$1,999.00', '$699.99', '$1,072.30', '$619.00', '$1,499.98',
       '$1,574.99', '$1,197.00', '$999.00', '$1,679.00', '$1,624.92',
       '$1,249.99', '$891.82', '$1,283.95', '$4,499.00', '$682.47',
       '$2,349.00', '$4,69

###Preprocessing Operations for the Price Feature
- Droping all the missing entries where Price is not present. Since Price is a target feature, retrieving missing prices is not possible.
- Transforming format of entries from '\$1,379.90' to '1379.90'
- Since Price is given in Dollars, we convert it into Rupees.
- Let's assume 1$ = ₹82.

In [35]:
df.shape

(2820, 17)

In [36]:
# Removing missing entries where Price is not present.
df = df.dropna(subset=['Price']).reset_index(drop=True)

In [37]:
df.shape

(1851, 17)

In [38]:
df.index

RangeIndex(start=0, stop=1851, step=1)

In [39]:
# Transforming format of entries from '$1,379.90' to '1379.90'
df['Price'] = df['Price'].str.replace('$', '').str.replace(',', '').astype(float)

In [40]:
# Converting dollars to rupees
df['Price'] = df['Price'] * 82

In [41]:
df['Price'].head()

0     29519.18
1    177694.00
2     67076.00
3     57318.00
4    200899.18
Name: Price, dtype: float64

##Addressing Screen_Size features's outliers
- It is less likely for laptop to have screen size smaller than 10 inches
- We will try to retrieve as much information as we can from the Title section.
- Rest of the outliers will be handled with respect to product Price

In [42]:
screen_outliers = df[df.Screen_Size < 10]
screen_outliers.Title.values

array(['Alienware Aurora R11 Gaming Desktop 10th Gen Intel Core i7, 16GB Dual Channel HyperX Fury DDR4 XMP, NVIDIA GeForce RTX 3080, 1TB HDD + 512GB SSD, Killer Wi-Fi 6, Black (Latest Model)',
       'Alienware Aurora R10 Gaming Desktop - AMD Ryzen 9 5900, 32GB DDR4 RAM, 1TB SSD, NVIDIA GeForce RTX 3080 10GB GDDR6X, Windows 10 Home - Lunar Light',
       'Alienware Aurora R8 Desktop, 9th Gen Intel Core i7 9700, NVIDIA GeForce RTX 2070, 8GB GDDR6 (Overclockable), 256GB SSD Boot + 2TB 7200 RPM, 16GB RAM, AWAUR8-7787SLV-PUS',
       'Alienware Dell Aurora R10 Gaming Desktop, AMD Ryzen 9 3900, 32GB Dual Channel HyperX Fury DDR4 XMP, 1TB SSD, AMD Radeon RX 5700 XT 8GB GDDR6, Lunar Light',
       'Alienware Aurora R10 Gaming Desktop - AMD Ryzen 9 5900, 32GB DDR4 RAM, 1TB SSD, NVIDIA GeForce RTX 3080 10GB GDDR6X, Windows 10 Home - Lunar Light',
       'Alienware New Aurora R11 Gaming Desktop, Intel i7-10700KF, NVIDIA GeForce RTX 2080 Super 8GB GDDR6, 512GB SSD + 1TB SATA HDD, 16GB DDR4 XMP, W

###**Observations**
- It appears that screen size information can't be extracted from the Title.
- So, we will try to reassign values based on the product Price

In [43]:
screen_outliers[['Screen_Size', 'Price']]

Screen_Size      Price
897         1.00  294379.18
904         1.00  171790.00
906         0.01  254118.00
909         1.00  188599.18
910         1.00  171790.00
919         0.10  270518.00

In [44]:
df[df.Price > 200000][['Screen_Size', 'Price']]

Screen_Size      Price
4            17.3  200899.18
7            17.0  250018.00
10           17.3  356290.00
11           16.0  250018.00
13           17.3  270190.00
...           ...        ...
1468         15.6  286918.00
1562         15.6  245999.18
1591         14.0  200900.00
1600         17.0  303318.00
1782         15.0  229599.18

[157 rows x 2 columns]

###**Insights**
- It appears that all of these outliers belongs to high configuration laptops.
- So it will be logical to assign them bigger screen size like 17.

In [45]:
# Reassigning screen size
screen_outliers.Screen_Size = 17

C:\Program Files\KMSpico\temp\ipykernel_10308\432613683.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  screen_outliers.Screen_Size = 17


In [46]:
screen_outliers.Screen_Size

897    17
904    17
906    17
909    17
910    17
919    17
Name: Screen_Size, dtype: int64

In [47]:
# Merging screen_outlier with df
df.loc[screen_outliers.index, 'Screen_Size'] = screen_outliers.Screen_Size

In [48]:
df.Screen_Size.value_counts().sort_index()

Screen_Size
10.00      1
10.10     11
10.40      1
10.50      3
10.51      1
11.00      4
11.60    103
12.00     10
12.10      1
12.20      9
12.30      6
12.40     20
12.45      2
13.00     19
13.10     16
13.30     54
13.40     11
13.50     30
14.00    376
14.10     12
14.20      1
14.40      5
14.50      5
15.00     36
15.40      1
15.60    744
16.00     92
16.20      2
16.60      1
17.00     37
17.25      1
17.30    226
18.40      6
30.00      1
31.50      2
34.00      1
Name: count, dtype: int64

##Addressing CPU_Model's missing entries one more time
###Filtering Operations for the CPU_Model feature
- Identifying entries with no missing values
- Converting all CPU_Model entries to lower case entries
- Design function containing regular expression to filter out categories
- Replacing all the unique values into i3, i5, i7, i9, ryzen3, ryzen5, ryzen7, ryzen9, celeron, ultra3, ultra5, ultra7, other  

In [49]:
df.CPU_Model.isna().sum()

93

In [50]:
df['CPU_Model'].unique()

array(['Core i5', 'Intel Core i7-11700', 'Intel Core i5-1135G7',
       'Ryzen 5', 'Intel Core i9', 'Core i7 8750H', 'Intel Core i7',
       'Core i7', 'AMD Ryzen 9 5900X', 'Core i9', 'Corei5-10300H', nan,
       'Celeron N4020', 'Core i7 Family', 'Intel Core i5', 'Ryzen 7',
       'ARM 7100', 'Core i3', 'Core I7 11800H', 'Corei7-10750H',
       'AMD Ryzen 7', 'Core i5 Family', 'Corei7-9750H', 'Ryzen 5 5500U',
       'Core i7-10510U', 'Corei7-1065G7', 'Xeon Platinum 8160M',
       'Ryzen 9', 'Core I9 12900H', 'AMD Ryzen 5 5500U', 'Celeron',
       'Celeron N', 'AMD A Series', 'Celeron N4000', 'Apple M1',
       'MediaTek MT8183', 'Ryzen 3 3350U', 'Core i7 Extreme', 'AMD A6',
       'Core i5-3340s', '8032', 'Pentium', 'Core I3 1115G4',
       'Intel Core i3', 'Ryzen 7 5800U', 'Unknown', 'AMD Ryzen 5 5600X',
       'MediaTek_MT8127', 'AMD A4', 'Celeron 2955U',
       'Intel Core i7-1165G7', 'Celeron N3350', 'Celeron 3205U',
       'Ryzen 3 2300U', 'Celeron N3450', 'Core i5-10210U', 'AMD 

In [51]:
# Filtering entries where CPU_Model value is not missing
cpu_not_missing = df[~df['CPU_Model'].isna()]

In [52]:
cpu_not_missing.shape

(1758, 17)

In [53]:
# Converting all CPU_Model entries to lower case entries
cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.lower()

C:\Program Files\KMSpico\temp\ipykernel_10308\3704837426.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.lower()


In [54]:
cpu_not_missing['CPU_Model'].unique()

array(['core i5', 'intel core i7-11700', 'intel core i5-1135g7',
       'ryzen 5', 'intel core i9', 'core i7 8750h', 'intel core i7',
       'core i7', 'amd ryzen 9 5900x', 'core i9', 'corei5-10300h',
       'celeron n4020', 'core i7 family', 'intel core i5', 'ryzen 7',
       'arm 7100', 'core i3', 'core i7 11800h', 'corei7-10750h',
       'amd ryzen 7', 'core i5 family', 'corei7-9750h', 'ryzen 5 5500u',
       'core i7-10510u', 'corei7-1065g7', 'xeon platinum 8160m',
       'ryzen 9', 'core i9 12900h', 'amd ryzen 5 5500u', 'celeron',
       'celeron n', 'amd a series', 'celeron n4000', 'apple m1',
       'mediatek mt8183', 'ryzen 3 3350u', 'core i7 extreme', 'amd a6',
       'core i5-3340s', '8032', 'pentium', 'core i3 1115g4',
       'intel core i3', 'ryzen 7 5800u', 'unknown', 'amd ryzen 5 5600x',
       'mediatek_mt8127', 'amd a4', 'celeron 2955u',
       'intel core i7-1165g7', 'celeron n3350', 'celeron 3205u',
       'ryzen 3 2300u', 'celeron n3450', 'core i5-10210u', 'amd r ser

In [55]:
cpu_not_missing.CPU_Model.value_counts()

CPU_Model
core i7            280
core i5            159
core i7 family     100
celeron n4020       95
intel core i7       82
                  ... 
celeron 2955u        1
core i5-2300         1
mediatek_mt8127      1
core i5-3340s        1
core i5 8400t        1
Name: count, Length: 130, dtype: int64

In [56]:
# Applying Regular Expression
pattern_cpu_filter1 = r'(intel\s*i\d|core\s*i\d|ryzen\s*\d|radeon\s*r\d|celeron|ultra\s*\d)'
cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.extract(pattern_cpu_filter1)

C:\Program Files\KMSpico\temp\ipykernel_10308\2944175908.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.extract(pattern_cpu_filter1)


In [57]:
cpu_not_missing['CPU_Model'].value_counts()

CPU_Model
core i7    529
core i5    336
celeron    269
core i3    116
ryzen 7     81
core i9     71
ryzen 5     69
ryzen 3     40
ryzen 9     30
corei7      17
corei5       6
Name: count, dtype: int64

In [58]:
# This function will rearrange values (from core i5--> core i5, ryzen 7--> ryzen7)
cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.replace('core ', 'core').str.replace('ryzen ', 'ryzen')

C:\Program Files\KMSpico\temp\ipykernel_10308\3114587673.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.replace('core ', 'core').str.replace('ryzen ', 'ryzen')


In [59]:
cpu_not_missing['CPU_Model'].value_counts()

CPU_Model
corei7     546
corei5     342
celeron    269
corei3     116
ryzen7      81
corei9      71
ryzen5      69
ryzen3      40
ryzen9      30
Name: count, dtype: int64

In [60]:
cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.replace('core', '')

C:\Program Files\KMSpico\temp\ipykernel_10308\585882354.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_not_missing['CPU_Model'] = cpu_not_missing['CPU_Model'].str.replace('core', '')


In [61]:
cpu_not_missing['CPU_Model'].value_counts().sort_index()

CPU_Model
celeron    269
i3         116
i5         342
i7         546
i9          71
ryzen3      40
ryzen5      69
ryzen7      81
ryzen9      30
Name: count, dtype: int64

In [62]:
cpu_not_missing.shape

(1758, 17)

##Replacing all the missing values of CPU_Model feature
- Calculating mean price per CPU_Model category
- Deciding price range for each category
- Replacing missing values based on the price range

In [63]:
# Replacing original dataframe df with cpu_not_missing
indexes = cpu_not_missing.index
df.loc[indexes, 'CPU_Model'] = cpu_not_missing['CPU_Model']

In [64]:
df.CPU_Model.isna().sum()

287

In [65]:
df.groupby('CPU_Model')['Price'].mean()

CPU_Model
celeron     19413.228699
i3          38358.285172
i5          74572.358480
i7         126433.542747
i9         270487.001690
ryzen3      36464.190500
ryzen5      50199.354203
ryzen7      74182.018765
ryzen9     150186.471333
Name: Price, dtype: float64

### Looks like most of the laptop Prices maintains following correlation with CPU_Model
- i3 : <50000
- i5 : 50000 - 100000
- i7 : 100000 - 200000
- i9 : >250000

###We will only use 'Core i' values family to replace missing entries since it is most famous.


In [66]:
# Filtering out rows with missing CPU_Model values
cpu_model_missing = df[df.CPU_Model.isna()]

In [67]:
# Assigning values
for i in cpu_model_missing.index:
  if cpu_model_missing.Price[i] < 50000:
    cpu_model_missing.CPU_Model[i] = 'i3'
  elif cpu_model_missing.Price[i] >= 50000 and cpu_model_missing.Price[i] <= 100000:
    cpu_model_missing.CPU_Model[i] = 'i5'
  elif cpu_model_missing.Price[i] > 100000 and cpu_model_missing.Price[i] <= 250000:
    cpu_model_missing.CPU_Model[i] = 'i7'
  else:
    cpu_model_missing.CPU_Model[i] = 'i9'

C:\Program Files\KMSpico\temp\ipykernel_10308\751738060.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_model_missing.CPU_Model[i] = 'i3'
C:\Program Files\KMSpico\temp\ipykernel_10308\751738060.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_model_missing.CPU_Model[i] = 'i3'
C:\Program Files\KMSpico\temp\ipykernel_10308\751738060.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cpu_model_missing.CPU_Model[i] = 'i7'
C

In [68]:
# Replacing dataframes
df.loc[cpu_model_missing.index, 'CPU_Model'] = cpu_model_missing.CPU_Model

In [69]:
df.isna().sum()

Title                  0
Brand                  0
Screen_Size            0
CPU_Model              0
Ram                   82
Operating_System      69
Rating               308
Series               208
Color                366
Resolution          1756
Weight              1823
USB                 1835
Battery             1848
Graphics             503
Disk_size            381
Card_desc            128
Price                  0
dtype: int64

##**Addressing RAM feature**
- RAM values will be extracted from the Title feature using Regular Expressions
- Removing outliers
- All the values will be converted into single unit(GB) with following formula
- 1 GB = 1024 MB, 1 TB = 1024 GB

In [70]:
df.Ram.unique()

array(['4 GB', '32 GB', '16 GB', '64 GB', nan, '8 GB', '128 GB', '12 GB',
       '20 GB', '24 GB', '2 GB', '6 GB', '1 TB', '40 GB', '16 MB'],
      dtype=object)

###It is highly unlikely for RAM to of size of 1TB

In [71]:
df[df.Ram == '1 TB']

Title      Brand  \
452   CF-31/MK6/16GB RAM/1TB SSD/Windows/Intel Core ...  Panasonic   
1825  SAMSUNG 15.6” Galaxy Book2 Pro with Intel ARC ...    SAMSUNG   

      Screen_Size CPU_Model   Ram Operating_System  Rating            Series  \
452          13.1        i5  1 TB       Windows 10     NaN            CF-318   
1825         15.6        i7  1 TB          Windows     4.1  Galaxy Book2 Pro   

             Color Resolution Weight  USB Battery         Graphics Disk_size  \
452   SILVER BLACK        NaN    NaN  NaN     NaN            Intel   1000 GB   
1825      Graphite        NaN    NaN  NaN     NaN  Intel ARC A350M       NaN   

       Card_desc      Price  
452   Integrated  377199.18  
1825   Dedicated  147599.18

In [72]:
ram_missing = df[df.Ram.isna()]
ram_missing.shape

(82, 17)

In [73]:
ram_missing.Title.values

array(['Ficihp 13.3" Laptop Screen Extender for M1 MacBook One Cable Plug&Play, Cooling System Triple Portable Monitor for Laptop with USB C HDMI, Monitor Attachment for Laptop Frame 13-16 inch Mac/Android',
       'B BELK Compatible with MacBook Air 13 inch Case M1, MacBook Air Case 2021 2020 2019 2018 Model A2337 A2179 A1932 with Touch ID, Clear Plastic Laptop Hard Shell + 2 Keyboard Covers + Screen Protector',
       'ASUS 17 Chromebook 17.3 Inch FHD Laptop 2023 Newest, Intel Celeron N4500 Up to 2.8Ghz, 4GBRAM, 192GB Storage, USB C, Wifi6, Bluetooth, 17hours Battery Life, Chrome OS',
       'Acer 2022 Aspire 5 Slim Laptop, 15.6" Full HD Display, AMD Ryzen 5 5500U Hexa Core Processor, AMD Radeon Graphics, WiFi 6, Backlit Keyboard, Windows 11 Home (24GB RAM | 1TB SSD)',
       'Acer Spin 7, 14" FHD Touch Display, Qualcomm Snapdragon 8cx Gen 2 5G Compute Platform, 8GB LPDDR4X, 512GB Universal Flash Storage, 5G, 4G LTE, Wi-Fi, Backlit KB, FPR, Antimicrobial, SP714-61NA-S1QA',
       'LG

In [74]:
ram_patterns = r'(\d{1,2}\s*GB)'
ram_missing['Ram'] = ram_missing.Title.str.extract(ram_patterns)

C:\Program Files\KMSpico\temp\ipykernel_10308\2396449359.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ram_missing['Ram'] = ram_missing.Title.str.extract(ram_patterns)


In [75]:
ram_missing.Ram.unique()

array([nan, '4GB', '24GB', '8GB', '16GB', '12GB', '32GB', '64GB'],
      dtype=object)

In [76]:
df.fillna(ram_missing, inplace=True)

In [77]:
df.Ram.isna().sum()

41

In [78]:
ram_missing[ram_missing['Ram'].isna()].Title.values

array(['Ficihp 13.3" Laptop Screen Extender for M1 MacBook One Cable Plug&Play, Cooling System Triple Portable Monitor for Laptop with USB C HDMI, Monitor Attachment for Laptop Frame 13-16 inch Mac/Android',
       'B BELK Compatible with MacBook Air 13 inch Case M1, MacBook Air Case 2021 2020 2019 2018 Model A2337 A2179 A1932 with Touch ID, Clear Plastic Laptop Hard Shell + 2 Keyboard Covers + Screen Protector',
       'LG FHD 32-Inch Computer Monitor 32ML600M-B, IPS with HDR 10 Compatibility, Black',
       'LG FHD 32-Inch Computer Monitor 32ML600M-B, IPS with HDR 10 Compatibility, Black',
       'Toughbook CF-31JEGAX1M 13.1" Notebook - Core i5 i5-2520M 2.50 GHz - Magnesium Alloy',
       'Panasonic Toughbook CF-31ACAAA1M 13.1" LED Notebook - Core i5 i5-520M 2.40 GHz',
       'Panasonic Toughbook CF-31JF7991M 13.1" LED Notebook - Intel Core i5 i5-2520M 2.50 GHz - 1024 x 768',
       'Sceptre 30-inch Curved Gaming Monitor 21:9 2560x1080 Ultra Wide/ Slim HDMI DisplayPort up to 200Hz Bu

###It appears that ram value is missing from the title description of the remaining rows. Since, we don't have any other way to find the ram value, we will drop all the rows with ram value missing.

In [79]:
df.shape

(1851, 17)

In [80]:
df.dropna(subset=['Ram'], inplace=True)

In [81]:
df.shape

(1810, 17)

In [82]:
df.isna().sum()

Title                  0
Brand                  0
Screen_Size            0
CPU_Model              0
Ram                    0
Operating_System      34
Rating               303
Series               174
Color                338
Resolution          1723
Weight              1784
USB                 1794
Battery             1807
Graphics             467
Disk_size            345
Card_desc             93
Price                  0
dtype: int64

In [83]:
df.Ram.unique()

array(['4 GB', '32 GB', '16 GB', '64 GB', '8 GB', '128 GB', '12 GB',
       '20 GB', '24 GB', '4GB', '24GB', '2 GB', '8GB', '6 GB', '1 TB',
       '16GB', '12GB', '40 GB', '32GB', '64GB', '16 MB'], dtype=object)

There is only one unique value for entry in TB and MB

##**Observations**
- It appears that unique entries have different units for the volume
- For model training, all the valumes must have same unit
- We will go with GB

In [84]:
def ram_to_number(string):
  """
  Extracts number from a string and convert it into number.

  Args:
    string : string containing magnetude and unit.

  Returns:
    A number in GB.
  """

  pattern_num = r'(\d{1,3})'  # Regular expression to identify number
  pattern_unit = r'(GB|MB|TB)'  # Regular expression to identify unit
  number = int(re.findall(pattern_num, string)[0])
  unit = re.findall(pattern_unit, string)[0]
  if unit == 'GB':
    return number
  elif unit == 'MB':
    return number / 1024
  else:
    return number * 1024

df.Ram = df.Ram.apply(ram_to_number)

In [85]:
df.Ram.value_counts().sort_index()

Ram
0.015625         1
2.000000        12
4.000000       347
6.000000        10
8.000000       472
12.000000       36
16.000000      615
20.000000       27
24.000000        6
32.000000      219
40.000000        3
64.000000       53
128.000000       7
1024.000000      2
Name: count, dtype: int64

##**Addressing Operating System feature**

In [86]:
df.Operating_System.unique()

array(['Windows 11 Pro', 'Windows 11 Home', 'Windows 10 Pro',
       'Windows 11', 'Windows 10 Home', 'Windows 7', 'Windows 10',
       'MacOS 12 Monterey', 'Win 10 PRO /Free upgrade to Win 11',
       'Windows 11 Home (S Mode)', 'Microsoft Windows 11 Professional',
       'Microsoft Windows 10 Professional (64-Bit)',
       'Microsoft Windows 10 Professional (64bit)',
       'Win 10 Pro / Free upgrade to Win 11',
       'Win 10 Multi-language /Free upgrade to Win 11',
       'Microsoft Windows 11 Home', 'Chrome OS', 'Windows 10 S',
       'Windows', 'Windows 11 Home in S Mode', 'Windows 11 S',
       'Windows 7 Professional', nan, 'Pc', 'Windows xp',
       'Windows 8 Professional', 'Windows 7 Professional, Windows 7',
       'Windows 8 Pro, Windows 7', 'Windows 11 Home S Mode', 'Android 10',
       'Window', 'Win 10 PRO Free upgrade to Win 11',
       'Windows 10 Home, Windows 11, Windows 10', 'Windows 11 in S mode',
       'Mac OS', 'Windows 8', 'Microsoft Windows 10 (64-Bit)',
    

In [87]:
df.Operating_System.isna().sum()

34

##**Observations**
###Maority of unique entries contains 'Windows' versions.
###Other entries got 'Mac OS' and 'Chrome OS'

##**Preprocessing**
###In order to check whether Operating Systems's significance with Price and handling missing values, following preprocessing operations should be performed
- Transform all values into lower case
- Filter out entries with no missing values and store it in a seperate dataframe. Continue further operations on new dataframe.
- Remove extra part from each string other than Operating System's name and its version. (like 'Windows 11 Home | Laptop Only' to 'windows 10')
- Calculate group mean of each Operating System's version with the Price and check whether it gives any insight.
- Based on the insights, apply modifications if required and merge new dataframe back with original.

In [88]:
# Filtering out entries with no missing values for Operating System feature
os_present = df[~df.Operating_System.isna()]

In [89]:
os_present.Operating_System.isna().sum()

0

In [90]:
os_present.shape

(1776, 17)

In [91]:
# Transforming entries into lower case
os_present.Operating_System = os_present.Operating_System.str.lower()

C:\Program Files\KMSpico\temp\ipykernel_10308\396848333.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  os_present.Operating_System = os_present.Operating_System.str.lower()


In [92]:
pattern_os = r'(\w+\s*\d*)' # Regular Expression to filter OS and it's version
os_present['Operating_System'] = os_present.Operating_System.str.extract(pattern_os)
os_present.Operating_System.value_counts()

C:\Program Files\KMSpico\temp\ipykernel_10308\2771543150.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  os_present['Operating_System'] = os_present.Operating_System.str.extract(pattern_os)


Operating_System
windows 11    840
windows 10    680
chrome        175
windows 7      27
microsoft      15
windows        12
win 10         10
windows 8       6
macos 12        3
windows         2
mac             2
pc              1
android 10      1
window          1
google          1
Name: count, dtype: int64

###Unique entries can be combined further in following manner:
- chrome, android 10, google, pc : chrome_os
- microsoft, windows, win 10, window : windows 10
- macos 12, mac : mac_os

In [93]:
def os_renaming(string):
  """
  This function standardizes operating system names to a few common values.

  Args:
    string: The original operating system name.

  Returns:
    A standardized operating system name (chrome_os, windows 10, or mac_os),
    or None if the input doesn't match any known names.
  """
  string = string.strip() # Important step : removing white space characters around the string
  if string in ['chrome', 'android 10', 'google', 'pc']:
    return 'chrome_os'
  elif string in ['microsoft', 'windows', 'win 10', 'window']:
    return 'windows 10'
  elif string in ['macos 12', 'mac']:
    return 'mac_os'
  else :
    return string

os_present['Operating_System'] = os_present.Operating_System.apply(os_renaming)

C:\Program Files\KMSpico\temp\ipykernel_10308\2633631600.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  os_present['Operating_System'] = os_present.Operating_System.apply(os_renaming)


In [94]:
os_present.Operating_System.value_counts()

Operating_System
windows 11    840
windows 10    720
chrome_os     178
windows 7      27
windows 8       6
mac_os          5
Name: count, dtype: int64

###Applying One Way Annova to check significance of Operating System feature with Price

In [95]:
grouped_data = os_present.groupby('Operating_System')['Price'].apply(list)
fvalue, pvalue = stats.f_oneway(*grouped_data)
print(fvalue, pvalue)

46.67892181084525 1.9757686880864124e-45


In [96]:
if pvalue < 0.05:
  print('Reject the null hypothesis')
else:
  print('Accept the null hypothesis')

Reject the null hypothesis


In [97]:
os_present.groupby('Operating_System')['Price'].mean().sort_index()

Operating_System
chrome_os      18495.717303
mac_os        168452.764000
windows 10     93866.754139
windows 11     89875.756381
windows 7     191741.328889
windows 8     144857.920000
Name: Price, dtype: float64

###Since, p_value is smaller than 0.05, we know different OS have maintain their significance with Price.
###Let's check whether there is a mean difference between different versions os 'windows' operating system.
###If there is no mean difference, we will combine all windows versions into single one.

In [98]:
# Filtering out entries where OS is a version of 'windows'
temp_os = os_present[(os_present.Operating_System == 'windows 7') | (os_present.Operating_System == 'windows 8') | \
 (os_present.Operating_System == 'windows 10') | (os_present.Operating_System == 'windows 11')]

In [99]:
grouped_data = temp_os.groupby('Operating_System')['Price'].apply(list)
fvalue, pvalue = stats.f_oneway(*grouped_data)
print(fvalue, pvalue)

16.47918960043456 1.5115600417120242e-10


In [100]:
if pvalue < 0.05:
  print('Reject the null hypothesis')
else:
  print('Accept the null hypothesis')

Reject the null hypothesis


###It is proven than there is no mean difference between different versions of windows OS, so unique values of OS shouldn't be updated anymore
###Now, this new dataframe should be combined with original one to handle missing values

In [101]:
df.loc[os_present.index, 'Operating_System'] = os_present.Operating_System

In [102]:
df.Operating_System.value_counts()

Operating_System
windows 11    840
windows 10    720
chrome_os     178
windows 7      27
windows 8       6
mac_os          5
Name: count, dtype: int64

In [103]:
df.Operating_System.isna().sum()

34

In [104]:
# Filtering out missing entries from Operating System feature
missing_os = df[df.Operating_System.isna()]

In [105]:
missing_os.groupby('CPU_Model')['Price'].mean()

CPU_Model
i3         30215.987059
i5         67778.487692
i7        187548.486667
ryzen3     40999.180000
Name: Price, dtype: float64

##**Observations**
- Usually, higher Core i version number correlates with higher Price. Same is observed during CPU_Model preprocessing steps. But, in this case correlation is reversed.
- Let's check whether there is same correlation present with individual Operating System in the original dataframe

In [106]:
df[df.Operating_System == 'windows 7'][['CPU_Model', 'Price']].groupby('CPU_Model')['Price'].mean()

CPU_Model
i3        237799.180000
i5        194612.622667
i7        193884.818000
ryzen7     81179.180000
Name: Price, dtype: float64

In [107]:
df[df.Operating_System == 'windows 8'][['CPU_Model', 'Price']].groupby('CPU_Model')['Price'].mean()

CPU_Model
i5    196799.18
i7    118887.29
Name: Price, dtype: float64

In [108]:
df[df.Operating_System == 'windows 10'][['CPU_Model', 'Price']].groupby('CPU_Model')['Price'].mean()

CPU_Model
celeron     21559.358812
i3          33927.627556
i5          75960.279351
i7         129968.529964
i9         323880.292667
ryzen3      33473.657333
ryzen5      50044.513684
ryzen7      76252.751200
ryzen9     167106.756364
Name: Price, dtype: float64

In [109]:
df[df.Operating_System == 'windows 11'][['CPU_Model', 'Price']].groupby('CPU_Model')['Price'].mean()

CPU_Model
celeron     21894.675294
i3          35127.235229
i5          61718.204343
i7         124181.165714
i9         256985.091321
ryzen3      38144.315833
ryzen5      50258.193600
ryzen7      73113.555636
ryzen9     140390.516842
Name: Price, dtype: float64

##**Observations**
- It appears that Operating_System with subcategory 'windows 7' and 'windows 8' also are following reverse correlation.
- Let's check whether entries with these Operating_Systems are having any outliers in the original dataframe

In [110]:
df.Operating_System.value_counts()

Operating_System
windows 11    840
windows 10    720
chrome_os     178
windows 7      27
windows 8       6
mac_os          5
Name: count, dtype: int64

In [111]:
df[df.Operating_System == 'windows 7'][['Brand', 'Ram', 'Disk_size', 'CPU_Model', 'Price']].sort_values('Price')

Brand   Ram Disk_size CPU_Model      Price
308   Toughbook   4.0    320 GB        i5   49199.18
400   Panasonic   2.0    160 GB        i5   65599.18
305   Panasonic   4.0    500 GB        i5   73799.18
1275       ASUS   4.0      1 TB    ryzen7   81179.18
438   Panasonic   4.0    500 GB        i5  114799.18
459   Toughbook   4.0    500 GB        i7  131199.18
15    Panasonic   4.0    320 GB        i5  159899.18
872   Alienware   8.0   1024 GB        i7  163918.00
891   Alienware   8.0      1 TB        i7  163984.42
864   Alienware   8.0      1 TB        i7  163984.42
874   Alienware  16.0   1024 GB        i7  163995.90
920   Alienware   8.0   1024 GB        i7  163995.90
441   Panasonic   4.0    320 GB        i5  163999.18
453   Panasonic   4.0    500 GB        i7  200572.00
489   Toughbook   4.0    320 GB        i5  200899.18
439   Panasonic   4.0    500 GB        i5  221399.18
435   Panasonic   4.0    320 GB        i7  229600.00
461   Panasonic   2.0    320 GB        i3  237799.18
436   Toughbook   4.0    500 GB        i5  241900.00
262   Toughbook   4.0       NaN        i5  245999.18
460   Panasonic   8.0    500 GB        i7  245999.18
449   Panasonic   4.0    500 GB        i5  250099.18
273   Panasonic   4.0    500 GB        i5  270599.18
454   Panasonic   4.0       NaN        i5  270600.00
437   Panasonic   4.0    500 GB        i5  286999.18
468   Panasonic   4.0    500 GB        i5  303399.18
306   Toughbook   4.0    500 GB        i7  311599.18

###**Observations**
- Hard to filter considerable number of outliers
- Higher Prices have higher versions of CPU_Models
- For a time being, we will assign following structure of missing values of Operating System
- For Price < 30000 : chrome_os
- For Price >= 30000 : windows 10

In [112]:
def os_assigner(price):
  """Assigns an operating system based on the price.

  Args:
    price: The price of the device.

  Returns:
    'chrome_os' if the price is less than 30000,
    'windows 10' otherwise.
  """
  if price < 30000:
    return 'chrome_os'  # Assign 'chrome_os' for lower-priced devices
  else:
    return 'windows 10'  # Assign 'windows 10' for higher-priced devices

missing_os['Operating_System'] = missing_os.Price.apply(os_assigner)

C:\Program Files\KMSpico\temp\ipykernel_10308\2491107593.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_os['Operating_System'] = missing_os.Price.apply(os_assigner)


In [113]:
missing_os.Operating_System.isna().sum()

0

In [114]:
# Merging with original dataframe
df.loc[missing_os.index, 'Operating_System'] = missing_os.Operating_System

In [115]:
df.isna().sum()

Title                  0
Brand                  0
Screen_Size            0
CPU_Model              0
Ram                    0
Operating_System       0
Rating               303
Series               174
Color                338
Resolution          1723
Weight              1784
USB                 1794
Battery             1807
Graphics             467
Disk_size            345
Card_desc             93
Price                  0
dtype: int64

##**Addressing Ratings**

In [116]:
df.Rating.unique()

array([3.9, 4.4, 4.3, 4.5, 5. , 4.1, nan, 3.1, 4.2, 4.8, 3.8, 3. , 4.9,
       2.8, 4.6, 3.7, 4. , 2.5, 4.7, 2. , 1. , 3.3, 3.6, 3.4, 3.5, 2.1,
       2.6, 3.2, 2.4, 2.7, 2.9, 2.3])

In [117]:
df.Rating.isna().sum()

303

- Retrieving rating through from some other feature like Title is not possible since Title doesn't describe ratings.
- Two options: Assign average ratings or assign zero ratings.
- We will asign zeo ratings to all missing entries.

In [118]:
df.Rating.fillna(0, inplace=True)

In [119]:
df.isna().sum()

Title                  0
Brand                  0
Screen_Size            0
CPU_Model              0
Ram                    0
Operating_System       0
Rating                 0
Series               174
Color                338
Resolution          1723
Weight              1784
USB                 1794
Battery             1807
Graphics             467
Disk_size            345
Card_desc             93
Price                  0
dtype: int64

###Features including Resolution, Weight, USB, Battery have missing entries almost equal to total entries. And from the Title description, it's hard to retrieve even 50% of those values. Hence, we will drop them.

##**Addressing Resolution, Weight, USB, Battery**

In [120]:
df.drop(['Resolution', 'Weight', 'USB', 'Battery'], axis=1, inplace=True)

In [121]:
df.isna().sum()

Title                 0
Brand                 0
Screen_Size           0
CPU_Model             0
Ram                   0
Operating_System      0
Rating                0
Series              174
Color               338
Graphics            467
Disk_size           345
Card_desc            93
Price                 0
dtype: int64

##**Addressing Series**

In [122]:
series = df.Series.str.lower().value_counts()
series

Series
inspiron                  34
pavilion                  23
aspire 5                  23
vivobook                  22
dell inspiron             21
                          ..
aero 17                    1
victus 15                  1
aorus 15p yd-73us244sh     1
antique                    1
ms1zo00001ob               1
Name: count, Length: 915, dtype: int64

In [123]:
series[series > 10]

Series
inspiron                   34
pavilion                   23
aspire 5                   23
vivobook                   22
dell inspiron              21
hp                         19
asus vivobook              16
hp pavilion                14
surface laptop             13
galaxy book pro            13
lg gram                    11
chromebook 4               11
dell alienware area 51m    11
Name: count, dtype: int64

In [124]:
df.shape

(1810, 13)

In [125]:
1810/2

905.0

###We have more than 50% of total entries as unique entries inside Series component. Most of them have occured only once and maximum count entry is of 34.

In [126]:
df.drop('Series', axis=1, inplace=True)

In [127]:
df.isna().sum()

Title                 0
Brand                 0
Screen_Size           0
CPU_Model             0
Ram                   0
Operating_System      0
Rating                0
Color               338
Graphics            467
Disk_size           345
Card_desc            93
Price                 0
dtype: int64

##**Addressing Color**

In [128]:
colors = df.Color.value_counts()
colors

Color
Black                 332
Silver                262
Gray                   66
Blue                   47
16GB RAM               42
                     ... 
Black / Silver Top      1
Red Backlit             1
Snowflake white         1
SPRUCE BLUE             1
Lunar light             1
Name: count, Length: 194, dtype: int64

In [129]:
df.Color.isna().sum()

338

In [130]:
colors[colors > 10]

Color
Black              332
Silver             262
Gray                66
Blue                47
16GB RAM            42
Natural Silver      39
32GB RAM            32
White               23
Carbon Gray         22
Core Black          21
Graphite            20
Grey                20
Platinum            20
Star Black          20
Rose Gold           17
64GB RAM            17
Gold                16
Slate Grey          15
Epic Silver         12
Platinum Silver     12
Pine Grey           11
Pure Silver         11
Name: count, dtype: int64

- Since we have 194 unique categories, and only 2 of them have entries more than 100.
- Rather than having multiple unique categories, we will select most occuring colors and do reassignment.
- All the missing entries will be replaced with 'other'

In [131]:
# Handling missing values for Color
df.Color.fillna('other', inplace=True)

In [132]:
def color_assigner(string):
  string = string.lower().strip()
  if 'black' in string:
    return 'black'
  elif 'white' in string:
    return 'white'
  elif 'silver' in string:
    return 'silver'
  elif 'gray' in string:
    return 'gray'
  elif 'gold' in string:
    return 'gold'
  elif 'blue' in string:
    return 'blue'
  elif 'blue' in string:
    return 'blue'
  else:
    return 'other'

df.Color = df.Color.apply(color_assigner)

In [133]:
df.Color.value_counts()

Color
other     666
black     472
silver    367
gray      125
blue       95
white      47
gold       38
Name: count, dtype: int64

In [134]:
df.isna().sum()

Title                 0
Brand                 0
Screen_Size           0
CPU_Model             0
Ram                   0
Operating_System      0
Rating                0
Color                 0
Graphics            467
Disk_size           345
Card_desc            93
Price                 0
dtype: int64

##**Addressing Graphics**

In [135]:
df.Graphics.isna().sum()

467

In [136]:
df.Graphics.unique()

array([nan, 'Intel UHD Graphics', 'NVIDIA GeForce GTX 1070',
       'NVIDIA GeForce RTX 3070', 'NVIDIA GeForce RTX 3070 Ti',
       'Intel Iris Xe Graphics', 'NVIDIA GeForce RTX 3080',
       'NVIDIA GeForce RTX 3060', 'Intel HD Graphics 4000',
       'NVIDIA GeForce GTX 1650', 'NVIDIA Quadro RTX A3000',
       'NVIDIA GeForce RTX 3080 Ti', 'NVIDIA Quadro RTX A2000',
       'NVIDIA GeForce GTX 1650 Max-Q', 'NVIDIA GeForce RTX 3050',
       'NVIDIA Quadro A2000', 'AMD Radeon RX6700M',
       'NVIDIA GeForce RTX 2070', '\u200eIntel Iris Xe', 'M1 Pro',
       'NVIDIA GeForce RTX 3050 Ti',
       'NVIDIA GeForce RTX 3060 Max-Q 6GB GDDR6', 'Intel UMA',
       'NVIDIA GeForce GTX 1050 Ti', 'AMD Radeon RX 6600',
       'NVIDIA GTX 1650 Ti MAX-Q', 'NVIDIA GeForce GTX 1660Ti',
       'NVIDIA GeForce RTX 2060', 'NVIDIA GeForce RTX 2080', 'Intel Iris',
       'NVIDIA GeForce GTX 1050', 'NVIDIA RTX A3000',
       'NVIDIA GTX 1650 MAX-Q', 'NVIDIA GeForce RTX 2080 Super',
       'NVIDIA GeForce GTX'

###**Observations and Modifications**
- We have three major graphics category belonging to Nvidia, AMD and Intel.
- We will merge all other categories into 'other'
- All the missing values can be retrieved from the Title description

 **Steps**

- Filter out all missing entries of Graphics feature and have a new dataframe
- Assign them values retrieved from Title feature (Nvidia, AMD, Intel)
- Assign 'other' for values that couldn't retrieve anything from Title description
- Merge modified dataframe with the original dataframe  
- Project non-missing entries on (Nvidia, AMD, Intel).

In [137]:
graphics_missing = df[df.Graphics.isna()]

In [138]:
graphics_missing.Title.values[:10]

array(['2022 Microsoft Surface Laptop Go 12.4" Touchscreen, Intel Core i5-1035G1 Processor, Up to 13Hr Battery Life, WiFi, Webcam, Windows 11 Pro, Platinum Silver (4GB RAM | 256GB SSD)',
       'HP Envy 34” All-in-One Desktop, NVIDIA GeForce GTX 1650, 11th Gen Intel Core i7-11700 Processor, 32 GB RAM, 512 GB SSD, WUHD IPS Display, Windows 11 Home, Detachable Magnetic Camera (34-c0010, 2021)',
       'MSI Bravo 15 15.6" 144Hz Gaming Laptop: AMD Ryzen R5-5600H, RX 5500M, 16GB, 512GB NVMe SSD, Win11 (B5DD-243)',
       'MSI Latest GE76 Raider Gaming Laptop | 17.3" 144 Hz FHD Display | Intel 14-Core i9-12900H | 64GB DDR5 2TB SSD | NVIDIA RTX3060 | WiFi 6 | HDMI | RJ45 | Thunberbolt4 | Backlit KB RGB | DTS | Win11 Pro',
       'MSI Raider GE66 15.6" QHD 240Hz Gaming Laptop: Intel Core i7-12700H RTX 3070 Ti 32GB DDR5 1TB NVMe SSD, USB-Type C, Thunderbolt 4, Cooler Boost 5, Win11 Home: Titanium Blue 12UGS-238',
       'Packard Bell CloudBook 14.1 inch Laptop, Full HD Display, Windows 11 Home 

In [139]:
def graphics_retriever(string):
  """
  This function identifies the graphics card manufacturer from a string.

  Args:
    string: The input string containing the description of the graphics card.

  Returns:
    The name of the graphics card manufacturer ("Nvidia", "AMD", "Intel", or "Other").
  """
  string = string.lower()
  if 'nvidia' in string:
    return 'Nvidia'
  elif 'amd' in string or 'radeon' in string:
    return 'AMD'
  elif 'intel' in string:
    return 'Intel'
  else:
    return 'Other'

# Apply the graphics_retriever function to the 'Title' column of the graphics_missing DataFrame
# to extract the graphics card manufacturer and store it in the 'Graphics' column.
graphics_missing['Graphics'] = graphics_missing.Title.apply(graphics_retriever)

C:\Program Files\KMSpico\temp\ipykernel_10308\2460385536.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  graphics_missing['Graphics'] = graphics_missing.Title.apply(graphics_retriever)


In [140]:
graphics_missing.Graphics.value_counts()

Graphics
Intel     316
Nvidia     53
AMD        53
Other      45
Name: count, dtype: int64

In [141]:
# Merging graphics_missing with original dataframe
df.loc[graphics_missing.index, 'Graphics'] = graphics_missing.Graphics

In [142]:
def graphics_categorization(graphics_string):
  """
  Categorizes graphics cards based on their names.

  Args:
    graphics_string: The name of the graphics card as a string.

  Returns:
    The category of the graphics card ("Nvidia", "AMD", "Intel", or "Other").
  """
  graphics_string = graphics_string.lower()
  if 'nvidia' in graphics_string:
    return 'Nvidia'
  elif 'amd' in graphics_string or 'radeon' in graphics_string:
    return 'AMD'
  elif 'intel' in graphics_string:
    return 'Intel'
  else:
    return 'Other'

# Apply the categorization function to the "Graphics" column of the DataFrame.
df['Graphics'] = df['Graphics'].apply(graphics_categorization)

In [143]:
df.Graphics.value_counts()

Graphics
Intel     1000
Nvidia     516
AMD        205
Other       89
Name: count, dtype: int64

In [144]:
df.isna().sum()

Title                 0
Brand                 0
Screen_Size           0
CPU_Model             0
Ram                   0
Operating_System      0
Rating                0
Color                 0
Graphics              0
Disk_size           345
Card_desc            93
Price                 0
dtype: int64

##**Addressing Disk_size**

In [145]:
df.Disk_size.isna().sum()

345

In [146]:
df.Disk_size.unique()

array(['256 GB', nan, '512 GB', '2 TB', '1 TB', '3 TB', '4 TB', '320 GB',
       '128 GB', '32 GB', '1000 GB', '1000 TB', '64 GB', '64', '4 GB',
       '500 GB', '16 GB', '1 GB', '628 GB', '160 GB', '6 GB', '256',
       '512', '6 TB', '1.5 TB', '1512 GB', '1256 GB', '1024 GB',
       '1128 GB', '2000 GB', '0.01 GB', '80 GB', '384 GB', '1', '65 GB',
       '2', '128 TB'], dtype=object)

In [147]:
df[df.Disk_size == '1'].Title.values

array(['Dell XPS 15 9520 (Latest Model) Intel 12th Gen Core i9-12900HK (14-CORE) 2TB PCIe SSD 64GB DDR5 RAM FHD+ Non Touch RTX 3050Ti Win 11 Pro (Certified Refurb)',
       'ASUS ROG Zephyrus G15 Ultra Slim Gaming Laptop, 15.6” 165Hz QHD Display, GeForce RTX 3080, AMD Ryzen 9 5900HS, 16GB DDR4, 1TB PCIe NVMe SSD, Wi-Fi 6, Windows 10, Eclipse Gray, GA503QS-BS96Q',
       'SAMSUNG Chromebook Plus (WiFi + LTE Verizon) Chrome OS 12.2" HD Touchscreen Intel Celeron 3965Y 4GB RAM 32GB eMMC - XE525QBB-K01US (Stealth Silver)'],
      dtype=object)

In [148]:
df.Disk_size.value_counts().sum()

1465

In [149]:
# Converting all disk units and Title string into lower case
df.Disk_size = df.Disk_size.str.lower()
df.Title = df.Title.str.lower()

In [150]:
df.head()

Title      Brand  Screen_Size  \
0  2022 microsoft surface laptop go 12.4" touchsc...  Microsoft         12.4   
1  hp envy 34” all-in-one desktop, nvidia geforce...         HP         34.0   
2  dell latitude 3510 business laptop black (inte...       Dell         15.6   
3  msi bravo 15 15.6" 144hz gaming laptop: amd ry...        MSI         15.6   
4  msi latest ge76 raider gaming laptop | 17.3" 1...        MSI         17.3   

  CPU_Model   Ram Operating_System  Rating   Color Graphics Disk_size  \
0        i5   4.0       windows 11     3.9  silver    Intel    256 gb   
1        i7  32.0       windows 11     4.4   other   Nvidia       NaN   
2        i5  16.0       windows 10     4.3   black    Intel    256 gb   
3    ryzen5  16.0       windows 11     4.5   black      AMD    512 gb   
4        i9  64.0       windows 11     5.0    blue   Nvidia      2 tb   

    Card_desc      Price  
0  Integrated   29519.18  
1         NaN  177694.00  
2  Integrated   67076.00  
3    RX 5500M   57318.00  
4   Dedicated  200899.18

In [151]:
disk_missing = df[df.Disk_size.isna()]

In [152]:
disk_missing.Title.values

array(['hp envy 34” all-in-one desktop, nvidia geforce gtx 1650, 11th gen intel core i7-11700 processor, 32 gb ram, 512 gb ssd, wuhd ips display, windows 11 home, detachable magnetic camera (34-c0010, 2021)',
       'excaliberpc 2022 msi creator z17 a12ugst-049 (i7-12700h, 32gb ram, 1tb nvme ssd, rtx 3070ti 8gb, 17" qhd+ 165hz touch, windows 11 pro) content creation laptop',
       'excaliberpc 2022 msi raider ge67hx 12ugs-013 (i7-12800hx, 16gb ram, 1tb nvme ssd, rtx 3070ti 8gb, 15.6" qhd 240hz, windows 11 pro) gaming laptop',
       'excaliberpc 2022 msi summit e13flipevo a12mt-013 (i7-1260p, 32gb ram, 1tb nvme ssd, intel iris xe, 13.4" fhd+ touch, windows 11 pro) professional laptop',
       'excaliberpc 2022 msi creator z16p b12uht-041 (i7-12700h, 32gb ram, 1tb nvme ssd, rtx 3080 8gb, 16" qhd+ 165hz touch, windows 11 pro) content creation laptop',
       'excaliberpc 2021 msi we76 11uk-460 (i7-11800h, 32gb ram, 1tb nvme ssd, quadro rtx a3000 6gb, 17.3" 144hz fhd, windows 10 pro) mob

###**Observations and Modifications**
###Most of the disk volumes for the missing entries is available inside the Title section and should be extracted from there using regular expressions.
**Steps:-**

- Filter out entries with missing disk volume and store it inside new dataframe
- Write a function to extract disk volumes from Title feature using regular expression
- Assign 512gb for the entries which still have missing entries
- Merge an updated dataframe with the original one.

In [153]:
# Regular expression to filter disk size with gb and tb as a unit
disk_patterns = r'(\d{1,3}\s*gb|\d{1}\s*tb)'

In [154]:
def disk_retriever(title, disk_patterns):
  """
  This function extracts disk size from a string using a regular expression.
  If multiple disk sizes are found, it returns the second one.

  Args:
    title: The string to extract the disk size from.
    disk_patterns: The regular expression to use for matching disk sizes.

  Returns:
    The extracted disk size or None if no disk size is found.
  """
  try:
    disks = re.findall(disk_patterns, title)
    if len(disks) > 1:
      return disks[1]
    else:
      return None
  except IndexError:
    return None

# Apply the disk_retriever function to the 'Title' column of the disk_missing DataFrame
# to extract the disk size and store it in the 'Disk_size' column.
disk_missing['Disk_size'] = disk_missing.Title.apply(lambda x: disk_retriever(x, disk_patterns))

C:\Program Files\KMSpico\temp\ipykernel_10308\1577772598.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  disk_missing['Disk_size'] = disk_missing.Title.apply(lambda x: disk_retriever(x, disk_patterns))


In [155]:
disk_missing.Disk_size.unique()

array(['512 gb', '1tb', '512gb', '256gb', '2tb', None, '128gb', '64gb',
       '64 gb', '16gb', '32gb', '160gb', '8gb', '4gb', '128 gb', '12 gb',
       '32 gb', '1 tb', '320 gb', '256 gb'], dtype=object)

In [156]:
disk_missing[disk_missing.Disk_size.isna()]['Title'].values

array(['panasonic toughbook cf-54g2878vm 256gb ssd, 8rg ram, emissive backlit keyboard, windows 10 pro',
       'panasonic toughbook laptop cf-31 sblax!m new!',
       'acer travelmate b3 tmb311-31-c343 - celeron n4020 / 1.1 ghz -',
       'hp victus 15.6" gaming laptop pc, nvidia geforce rtx 3050 ti, amd ryzen 7 5800h, refined 1080p ips display, compact design, all-in-one keyboard with enlarged touchpad, hd webcam (15-fb0028nr, 2022)',
       'lg 15zt90p-g.am33u1 15" gram lightweight notebook',
       'lg 17zt90p-g.ax33u1 17" gram lightweight notebook',
       '15.6 inch laptop,windows 10 pro, 1366×768 hd screen,intel celeron n5100 processor,6gb ram，128g ssd',
       'gigabyte [2020] aero 15s oled xb thin+light high performance laptop, 15.6" 4k uhd oled display w/ 100% dci-p3, geforce rtx 2070 super max-q, i7-10875h, 16gb ddr4',
       'alienware aw13r3-7000slv-pus 13.3" gaming laptop (7th generation intel core i7, 8gb ram, 256 ssd, silver) with nvidia gtx 1060',
       'samsung np340

In [157]:
# Merging updated dataframe with the original one
df.loc[disk_missing.index, 'Disk_size'] = disk_missing.Disk_size

In [158]:
df.Disk_size.isna().sum()

13

In [159]:
# Replacing remaining missing values with '512gb'
df.Disk_size.fillna('512gb', inplace=True)

- It appears that unique entries have different units for the volume
- For model training, all the valumes must have same unit
- We will go with GB


In [160]:
df.Disk_size.value_counts()

Disk_size
512 gb     364
1 tb       352
256 gb     218
64 gb      146
128 gb     128
512gb      108
1tb        103
32 gb       67
2 tb        57
256gb       33
1000 gb     30
128gb       24
64gb        21
16 gb       20
500 gb      19
32gb        16
320 gb      11
4 tb        11
16gb        11
512          6
2tb          6
1256 gb      5
8gb          5
160 gb       5
64           4
1024 gb      4
1128 gb      4
1512 gb      3
4gb          3
256          3
1            3
4 gb         2
1.5 tb       2
1 gb         2
2000 gb      1
2            1
65 gb        1
384 gb       1
80 gb        1
0.01 gb      1
628 gb       1
3 tb         1
12 gb        1
6 tb         1
1000 tb      1
160gb        1
6 gb         1
128 tb       1
Name: count, dtype: int64

In [161]:
def disk_to_number(string):
  """
  Extracts number from a string and convert it into number.

  Args:
    string : string containing magnetude and unit.

  Returns:
    A number in GB.
  """

  pattern_num = r'(\d{1,3})'  # Regular expression to identify number
  pattern_unit = r'(gb|mb|tb)'  # Regular expression to identify unit
  number = int(re.findall(pattern_num, string)[0])
  try:
    unit = re.findall(pattern_unit, string)[0]
  except IndexError:
    if number < 6:  # Assigning units for the volumes which don't have one
      unit = 'tb'
    else:
      unit = 'gb'

  if unit == 'gb':
    return number
  elif unit == 'mb':
    return number / 1024
  else:
    return number * 1024

df.Disk_size = df.Disk_size.apply(disk_to_number)

In [162]:
df.Disk_size.value_counts()

Disk_size
512       478
1024      460
256       254
64        171
128       152
32         83
2048       64
16         31
100        30
500        19
4096       11
320        11
160         6
125         5
8           5
4           5
112         4
102         4
151         3
1           2
80          1
384         1
0           1
200         1
65          1
628         1
12          1
6144        1
6           1
102400      1
3072        1
131072      1
Name: count, dtype: int64

###**Observations and Modifications**
- It appears that we have plenty of outliers present inside Disk_size
- Usual range for Disk volumes lies between 16gb to 4tb (16gb - 4096gb)
- Also, there are lot of entries present which doesn't look natural. Disk volumes are usually comes in power of 2.
- Let's find out those outliers first and store them in a seperate dataframe

In [163]:
# Filtering out outliers with volume less than 16gb and greater than 4tb
disk_outliers1 = df[(df.Disk_size < 16) | (df.Disk_size > 4096)]
disk_outliers1

Title        Brand  \
136   msi gf63 thin 9sc-066 15.6" gaming laptop, thi...          MSI   
246   2021 premium acer chromebook 15.6" hd laptop l...         Acer   
320   acer aspire 5 slim laptop, 15.6" full hd ips d...         Acer   
392   bocconi leadbook t6 laptop | 15.6" full hd 192...        WAKST   
413   fz-g1u/ea21/barcode scanner/mk5/core i5-7300u/...    Panasonic   
454   panasonic cf-31sal781m win7, intel core i5-332...    Panasonic   
560   gigabyte aero 17 hdr sa-7us4020sh 17.3" thin b...     GIGABYTE   
579   excaliberpc 2023 msi titan gt77hx 13vh-046us p...  EXCaliberPC   
643   fusion5 14.1" t90 n4 intel quad core windows 1...      Fusion5   
711   laptop 15-dy2097nr|intel core i7 11th gen|wind...      Generic   
1000  hp pavilion x360 15.6 inch 2-in-1 laptop pc, 1...           HP   
1194  samsung galaxy book2 pro 13.3” 256gb laptop co...      SAMSUNG   
1309  asus laptop l406 thin and light laptop, 14” hd...         ASUS   
1344  asus chromebook vibe cx34 flip cloud gaming la...         ASUS   
1444  15.6" s15 n2 full hd windows 11 professional s...      Fusion5   
1523  2020 newest dell inspiron 11 3195 2-in-1 11.6 ...         Dell   
1544  samsung galaxy chromebook enterprise (256gb st...      SAMSUNG   
1777  thomson laptop neo 10, 10.1 inch, intel atom ,...      Thomson   

      Screen_Size CPU_Model    Ram Operating_System  Rating   Color Graphics  \
136          15.6        i7   16.0       windows 10     4.5   black   Nvidia   
246          15.6   celeron    4.0        chrome_os     4.6  silver    Intel   
320          15.6    ryzen7   16.0       windows 11     5.0   other      AMD   
392          15.6   celeron    6.0       windows 10     3.5    gray    Other   
413          10.1        i7    8.0       windows 10     0.0   other    Intel   
454          13.0        i5    4.0        windows 7     0.0   other    Intel   
560          17.0        i7    6.0       windows 10     4.0   other   Nvidia   
579          17.3        i9  128.0       windows 11     0.0   other   Nvidia   
643          14.0        i3    4.0       windows 11     4.4   other    Intel   
711          15.0        i5   12.0       windows 11     0.0   other    Intel   
1000         15.6        i5   12.0       windows 10     4.5  silver    Intel   
1194         13.3        i5    8.0       windows 11     3.8  silver    Intel   
1309         14.0   celeron    4.0       windows 10     4.3    gray    Intel   
1344         14.0        i5    8.0        chrome_os     3.0   white   Nvidia   
1444         15.6   celeron    4.0       windows 11     3.9   other    Intel   
1523         11.6        i3    4.0       windows 10     4.1    gray      AMD   
1544         13.3        i5    8.0        chrome_os     3.2   other    Intel   
1777         10.1        i3    4.0       windows 10     3.2   white    Intel   

      Disk_size   Card_desc      Price  
136      102400   Dedicated   77818.00  
246           4  Integrated   18778.00  
320           1   Dedicated   56579.18  
392           6  Integrated   14759.18  
413           8  Integrated  204999.18  
454           4  Integrated  270600.00  
560           8   Dedicated  163918.00  
579        6144         NaN  499298.00  
643           4  Integrated   20499.18  
711          12         NaN   57070.36  
1000          0  Integrated   57318.00  
1194          8  Integrated   90199.18  
1309          1  Integrated   16318.00  
1344          8  Integrated   63139.18  
1444          4  Integrated   24599.18  
1523     131072  Integrated   19598.00  
1544          8  Integrated   98807.54  
1777          4  Integrated    8036.00

In [164]:
# Filtering out entries where disk size is not in power of 2
# Define a function to check if a number is a power of 2
def is_power_of_two(n):
  return (n & (n-1) == 0) and n != 0

# Apply the function to the 'Disk_size' column and filter for values that are not powers of 2
disk_outliers2 = df[~df['Disk_size'].apply(is_power_of_two)]
disk_outliers2

Title        Brand  \
10    excaliberpc 2022 msi raider ge76 12uh-655 pro ...  EXCaliberPC   
15    panasonic toughbook cf-53sjczylm laptop (windo...    Panasonic   
122   msi gl62m 7rex-1896us 15.6" full hd gaming lap...          MSI   
136   msi gf63 thin 9sc-066 15.6" gaming laptop, thi...          MSI   
154   msi gl62m 7rex-1896us 15.6" full hd gaming lap...          MSI   
...                                                 ...          ...   
1671  dell inspiron 3583 15€  laptop intel celeron...         Dell   
1689  2022 asus 14" thin light business student lapt...         ASUS   
1716  2022 newest hp laptop with microsoft office in...           HP   
1717  2022 asus 14" thin light business student lapt...         ASUS   
1744  hp 2022 newest touch-screen laptops for colleg...           HP   

      Screen_Size CPU_Model   Ram Operating_System  Rating   Color Graphics  \
10           17.3        i9  64.0       windows 11     5.0   other   Nvidia   
15           14.0        i5   4.0        windows 7     3.1  silver    Intel   
122          15.6        i7   8.0       windows 10     4.3   black   Nvidia   
136          15.6        i7  16.0       windows 10     4.5   black   Nvidia   
154          15.6        i7   8.0       windows 10     4.3   black   Nvidia   
...           ...       ...   ...              ...     ...     ...      ...   
1671         15.6   celeron   4.0       windows 10     4.3   black    Intel   
1689         14.0   celeron   4.0       windows 11     4.3    gold    Intel   
1716         15.6        i3  16.0       windows 11     4.2    gold      AMD   
1717         14.0   celeron   4.0       windows 11     4.3    gold    Intel   
1744         15.6        i5  16.0       windows 11     4.0  silver    Intel   

      Disk_size   Card_desc      Price  
10         3072   Dedicated  356290.00  
15          320  Integrated  159899.18  
122         100   Dedicated   81999.18  
136      102400   Dedicated   77818.00  
154         100   Dedicated   81999.18  
...         ...         ...        ...  
1671        500  Integrated   18445.90  
1689        320         NaN   20499.18  
1716        100  Integrated   45099.18  
1717        320         NaN   20499.18  
1744        100  Integrated   54939.18  

[93 rows x 12 columns]

In [165]:
# Combining disk_outliers1 and disk_outliers2
disk_outliers = pd.concat([disk_outliers1, disk_outliers2])
disk_outliers.shape

(111, 12)

In [166]:
disk_outliers.Title.values

array(['msi gf63 thin 9sc-066 15.6" gaming laptop, thin bezel, intel core i7-9750h, nvidia geforce gtx1650, 16gb, 512gb nvme ssd',
       '2021 premium acer chromebook 15.6" hd laptop light computer, intel celeron n4000, 4gb ram, 32gb emmc, hd webcam, intel uhd graphics 600, 12+ hours battey, sleeve, bluetooth, chrome os,w/marxsol cables',
       'acer aspire 5 slim laptop, 15.6" full hd ips display, amd ryzen7 5700u octa-core processor, amd radeon rx vega 8 graphics, 16gb ram, 1tb ssd, windows 11 home, fingerprint reader, twe hdmi cable',
       'bocconi leadbook t6 laptop | 15.6" full hd 1920x1080 ips display | intel celeron n3450 quad-core | 6gb ddr3 | 128gb ssd | ac wifi | mini hdmi | windows 10 home',
       'fz-g1u/ea21/barcode scanner/mk5/core i5-7300u/2.60ghz /win10 pro 64 bit/ 10.1" wuxga gloved multi touch + digitizer lcd /256gb ssd/8gb /wi-fi/bt/dual pass/rear cam/infrared hello webcam/tpm 2.0',
       'panasonic cf-31sal781m win7, intel core i5-3320m 2.60ghz, vpro, 13.1 xga

###**Insights**
- As in previous cases, we have Disk volume information available inside the Title section
- We will again use regular expressions for the retrieval

In [167]:
def disk_retrieval(string):
  disk_pattern = r'(\d{1,3}\s*gb|\d{1,3}\s*tb)'
  disk_values = re.findall(disk_pattern, string)

  # In some Title description, sequence of Ram and Disk is interchanged
  # So, we will select biggest volume as a disk volume

  values_in_gb = []
  for value in disk_values:
    num_pattern = r'(\d{1,3})'
    num = int(re.findall(num_pattern, value)[0])
    if 'tb' in value:
      values_in_gb.append(num * 1024)
    if 'gb' in value:
      values_in_gb.append(num)

  if len(values_in_gb) == 0:  # This will handle cases where no Ram or Disk value can be retrieved
    return 512
  return max(values_in_gb)

disk_outliers.Disk_size = disk_outliers.Title.apply(disk_retrieval)

In [168]:
# Merging disk_outliers with the original dataframe
df.loc[disk_outliers.index, 'Disk_size'] = disk_outliers.Disk_size

In [169]:
df.Disk_size.value_counts().sort_index()

Disk_size
4         2
16       32
32       84
64      175
80        1
128     156
160       4
256     262
320       9
500      14
512     486
1024    507
2048     65
3072      1
4096     11
6144      1
Name: count, dtype: int64

In [170]:
df[df.Disk_size == 500].Title.values

array(['panasonic toughbook cf 31 13.1" led intel core i5-5300u 2.3ghz 4gb ram 500gb hdd windows 7 professional rugged notebook cf-3110451cm',
       'tb 31 i7-2640m 2.80g 4gb 500gb 13.1in-xga wl tpm bt w7p dual pass',
       'acer 2022 newest aspire 5 slim 15.6" fhd laptop, 11th gen intel core i3-1115g4(up to 4.1ghz,beat i5-7200u), 8gb ddr4 ram, 128gb ssd+500gb hdd, wifi 6, usb-c, webcam, hdmi, windows 11s+jvq mp',
       'panasonic cf-54f0001vm semi rugged toughbook (intel core i5-6300u 2.40ghz, 500gb hard drive, 4gb ram, emissive backlit keyboard)',
       'toughbook cf-191myax1m 10.1" led intel core i5 i5-3320m 2.60ghz 4gb ram 500gb hdd win 7 pro 64-bit rugged notebook',
       'toughbook cf-31wflaxlm 13.1" touchscreen lcd led intel core i5 i5-3340m 2.70 ghz 4gb ram 500gb hdd windows 7 64-bit notebook',
       'panasonic toughbook cf-53stlzylm 14-inch laptop (2.7 ghz intel core i5 3320m processor, 4gb ddr3l, 500gb hdd, windows 8) silver',
       'panasonic cf-54e6001vm semi rugged 

In [171]:
df.isna().sum()

Title                0
Brand                0
Screen_Size          0
CPU_Model            0
Ram                  0
Operating_System     0
Rating               0
Color                0
Graphics             0
Disk_size            0
Card_desc           93
Price                0
dtype: int64

##**Addressing Card_desc**

In [172]:
card_count = df.Card_desc.value_counts()
card_count

Card_desc
Integrated                                                        1090
Dedicated                                                          481
RTX 3060                                                            12
RTX 3050 Ti                                                         12
Iris Xe Graphics                                                    11
Intel XE                                                             9
Integrated, Dedicated                                                8
RTX 3070                                                             8
RTX 3050                                                             7
Intel Iris Xe                                                        5
Intel UHD Graphics                                                   5
RTX 3080 Ti                                                          5
Intel Iris Xᵉ                                                        4
UHD Graphics 500, Integrated                                       

In [173]:
df.Card_desc.unique().shape

(52,)

###**Observations and Modifications**
- This Graphics card descriptions looks similar to Graphics feature.
- Out of 52 unique categories, only 2 of them carries more than 75% of values and have individual count of more than 100
- Rest of the entries have value count less than 15
- We will group all those entries along with missing one into the 'other' category

In [174]:
# Replacing all the null entries with 'other'
df.Card_desc.fillna('other', inplace=True)

In [175]:
def card_categorization(card_string):
  """
  Categorizes graphics cards based on their names.

  Args:
    card_string: The name of the graphics card as a string.

  Returns:
    The category of the graphics card ("Integrated", "Dedicated", or "Other").
    """
  card_string = card_string.lower()
  if 'integrated' in card_string:
    return 'Integrated'
  elif 'dedicated' in card_string:
    return 'Dedicated'
  else:
    return 'other'

# Apply the categorization function to the "Graphics" column of the DataFrame.
df['Card_desc'] = df['Card_desc'].apply(card_categorization)

In [176]:
df.isna().sum()

Title               0
Brand               0
Screen_Size         0
CPU_Model           0
Ram                 0
Operating_System    0
Rating              0
Color               0
Graphics            0
Disk_size           0
Card_desc           0
Price               0
dtype: int64

In [177]:
df.head()

Title      Brand  Screen_Size  \
0  2022 microsoft surface laptop go 12.4" touchsc...  Microsoft         12.4   
1  hp envy 34” all-in-one desktop, nvidia geforce...         HP         34.0   
2  dell latitude 3510 business laptop black (inte...       Dell         15.6   
3  msi bravo 15 15.6" 144hz gaming laptop: amd ry...        MSI         15.6   
4  msi latest ge76 raider gaming laptop | 17.3" 1...        MSI         17.3   

  CPU_Model   Ram Operating_System  Rating   Color Graphics  Disk_size  \
0        i5   4.0       windows 11     3.9  silver    Intel        256   
1        i7  32.0       windows 11     4.4   other   Nvidia        512   
2        i5  16.0       windows 10     4.3   black    Intel        256   
3    ryzen5  16.0       windows 11     4.5   black      AMD        512   
4        i9  64.0       windows 11     5.0    blue   Nvidia       2048   

    Card_desc      Price  
0  Integrated   29519.18  
1       other  177694.00  
2  Integrated   67076.00  
3       other   57318.00  
4   Dedicated  200899.18

###We have addressed all the missing entries, but still there is some preprocessing operations are left to do like outlier removal, which will be continued further in new notebook. For a time being, we will store this dataframe with name 'df_cleaned_24_09_2024.csv'

In [178]:
df.shape

(1810, 12)

In [179]:
# df.to_csv('df_cleaned_25_09_2024.csv')